service/history/historyReplicator.go

```diff
@@ -88,7 +89,42 @@ func (r *historyReplicator) ApplyEvents(request *h.ReplicateEventsRequest) (retE
 		if err != nil {
 			return err
 		}
+
+		// Check for out of order replication task and store it in the buffer
+		if firstEvent.GetEventId() > msBuilder.GetNextEventID() {
+			if t := msBuilder.BufferReplicationTask(request); t == nil {
+				return errors.New("failed to add buffered replication task")
+			}
+
+			return nil
+		}
+	}
+
+	// First check if there are events which needs to be flushed before applying the update
+	err = r.FlushBuffer(context, msBuilder, request)
```

Comment:
the flush buffer function will internally call ApplyReplicationTask, so is there any redundant logic here?

Reply:
Flush calls it for buffered replication tasks, not for the current one being processed.